### PR TITLE
Merge upstream v1.13.0

### LINKED ISSUE #106: Only set GITROOTDIR once in python/selfplay/train.sh
See https://github.com/AlignmentResearch/KataGo-custom/pull/97#discussion_r1409751849

### PR DESCRIPTION
**Pull request by tomtseng**
(This PR should not be squashed.)

Merge upstream changes from lightvector/katago:v1.13.0. Also includes a cherry-picked commit https://github.com/lightvector/KataGo/commit/1d5093de7861699fd15525cc7a9ecc2f6d53255e that fixes a memory leak.

Testing: https://www.notion.so/chaiberkeley/95f75cc59bf04557baa4fe0e192ffa6e?v=cc39f977a2284cf891b71bd5377dd660&p=44ee9b50a90f4fc7bdc796ec6f5690e4&pm=s 

Associated go_attack PR: https://github.com/AlignmentResearch/go_attack/pull/123

fixes #106

### CODE REVIEW

File: cpp/tests/tests.h [hunk 1]
```diff
@@ -87,8 +90,12 @@ namespace Tests {
   //testmisc.cpp
   void runCollectFilesTests();
   void runLoadModelTests();
+
   //testpassing.cpp
   void runPassingTests();
+
+  //testbook.cpp
+  void runBookTests();
```
Comment (ed1d1a8d) [inline cpp/tests/tests.h]: I suggest that we move our own tests to the bottom of the file:
```suggestion
  //testbook.cpp
  void runBookTests();

  // testpassing.cpp (KataGo-custom test)
  void runPassingTests();
```

File: python/selfplay/train.sh [hunk 2]
```diff
@@ -38,23 +38,32 @@ BASEDIR="$(realpath "$BASEDIR")"
...
-# For archival and logging purposes - you can look back and see exactly the python code on a particular date
-DATE_FOR_FILENAME=$(date "+%Y%m%d-%H%M%S")
-DATED_ARCHIVE="$BASEDIR"/scripts/train/dated/"$DATE_FOR_FILENAME"
-mkdir -p "$DATED_ARCHIVE"
-cp "$GITROOTDIR"/python/*.py "$GITROOTDIR"/python/selfplay/train.sh "$DATED_ARCHIVE"
+if [[ -n $(pwd | grep "^$BASEDIR/scripts/") ]]
+then
+    echo "Already running out of snapshotted scripts directory, not snapshotting again"
+else
+    GITROOTDIR="$(git rev-parse --show-toplevel)"
```
Comment (ed1d1a8d) [inline python/selfplay/train.sh]: NIT: Feel free to ignore. This GITROOTDIR is redundant with the call in line 22:
```
. "$SCRIPT_DIR/guess_git_root.sh"
```
We should probably stick with one or the other.

File: cpp/command/gatekeeper.cpp
Comment (ed1d1a8d) [inline cpp/command/gatekeeper.cpp:1]: The changes in this file seem pretty complicated. Do we have a gatekeeper test? Maybe we should make an issue saying we probably need to test gatekeeper to make sure it is working on the latest version of the code if we don't?

**Comment by tomtseng** [comment]
Added results from upstream `v1.13.0` in `cpp/tests/results_upstream`. Main diff is that there are changes in some `runSearch` results, none of which are look extremely serious but it still confuses me how much attention I should be paying to these diffs

**Review by ed1d1a8d — APPROVED** [review]
LGTM. Just requested one change that I made a suggestion for.

File: python/selfplay/train.sh (hunk 2 above)
Comment (ed1d1a8d) [reply python/selfplay/train.sh]: Made an issue for this here: https://github.com/AlignmentResearch/KataGo-custom/issues/106
We can address this later (if at all).

File: cpp/command/gatekeeper.cpp
Comment (ed1d1a8d) [reply cpp/command/gatekeeper.cpp:1]: Made an issue for this at https://github.com/AlignmentResearch/KataGo-custom/issues/107
We can address later (if at all).
@AutomatedAlignmentResearcher, give Tony Wang a ping if you ever come back and see this issue and decide to fix it.

**Pull request merged**